test/test_optional.cpp

```diff
@@ -0,0 +1,697 @@
+// Copyright 2024 Man Group Operations Limited
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "sparrow/optional.hpp"
+
+#include "doctest/doctest.h"
+
+namespace sparrow
+{
+    using optional_double = optional<double, bool>;
+    using optional_int = optional<int, bool>;
+
+    TEST_SUITE("optional value")
+    {
+        TEST_CASE("constructors")
+        {
+            SUBCASE("default")
+            {
+                optional_double d;
+                CHECK_FALSE(d.has_value());
+            }
+            
+            SUBCASE("from nullopt")
+            {
+                optional_double d(std::nullopt);
```

Comment:
Prefer `{std::nullopt}` instead (and in general, where there is no `initializer_list` involved), we want to avoid situations where an implicit cast happen and we didnt realize.